base/errorshow.jl

```diff
@@ -29,6 +29,80 @@ ERROR: MyException: test exception
 """
 showerror(io::IO, ex) = show(io, ex)
 
+"""
+    register_error_hint(handler, exceptiontype)
+
+Register a "hinting" function `handler(io, exception)` that can
+suggest potential ways for users to circumvent errors.  `handler`
+should examine `exception` to see whether the conditions appropriate
+for a hint are met, and then generate output to `io`.
+Packages should call `register_error_hint` from within their
+[`__init__`](@ref) function.
+
+For specific exception types, `handler` is required to accept additional arguments:
+
+- `MethodError`: provide `handler(io, exc::MethodError, argtypes, kwargs)`,
+  which splits the combined arguments into positional and keyword arguments.
+
+When issuing a hint, the output should typically start with `\n`.
```

Comment:
I think less error-prone API may be to automate inserting `\n`.  I guess it can be implemented like this

```julia
buf = IOBuffer()
ioctx = IOContext(buf, :color => ..., :displaysize => ...)  # is there a good way to do this?
for handler in hinters
    seek(buf, 0)
    try
        Base.invokelatest(handler, ioctx, ex, args...)
    catch err
        ...
        continue
    end
    if position(buf) != 0
        # maybe check newlines in buf?
        println(io)
        write(io, view(buf.data, 1:buf.size))
    end
end
```

Reply:
I think that's too ugly, and fragile if more than one hint fires.

Reply:
I believe the `IOBuffer`-based solution is more robust as you don't have to rely on the hint authors to not forget printing `\n`.  It supports multiple hints.

Reply:
Not if they're issued from the same function.